### PR TITLE
Bump Go version, dependency versions, add example function

### PR DESCRIPTION
**Pull request by bflad**
Reference: https://developer.hashicorp.com/terraform/plugin/framework/functions
Reference: https://github.com/hashicorp/terraform-providers-devex-internal/issues/175

Terraform 1.8 and later introduce the concept of provider-defined functions and this adds an example function that accepts a string argument and returns the same value as a string return.

Dependencies updated via:

```
go get github.com/hashicorp/terraform-plugin-framework@v1.6.0
go get github.com/hashicorp/terraform-plugin-docs@v0.18.0
go mod tidy
```

### CODE REVIEW

**Comment by bflad** [comment]
The `generate` tests are unfortunately not going to work until Terraform 1.8.0 final is released since `tfplugindocs` will remove the `docs/functions` directory to double check there are no changes on regeneration. 

```
Run git diff --compact-summary --exit-code || \
 docs/functions/example.md (gone) | 26 --------------------------
 1 file changed, 2[6](https://github.com/hashicorp/terraform-provider-scaffolding-framework/actions/runs/8080901492/job/22078273130?pr=198#step:5:7) deletions(-)

Unexpected difference in directories after code generation. Run 'go generate ./...' command and commit.
```

We would need to hardcode the `tfplugindocs generate` command invocation to include `-tf-version ...` or call the `hashicorp/setup-terraform` action beforehand with a 1.8 prerelease. I think the latter is preferable as a temporary workaround that we can code comment, but we need 1.8.0-beta1 to be released, which should happen later today.

File: internal/provider/example_function_test.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+					value = provider::scaffolding::example(terraform_data.test.output)
+				}
+				`,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckOutput("test", "testvalue"),
+				),
+			},
+		},
+	})
+}
```
Comment (austinvalle) [inline internal/provider/example_function_test.go:76]: ❤️

**Review by austinvalle — APPROVED** [review]
LGTM 🚀 :shipit:

**Review by bendbennett — APPROVED** [review]
LGTM 🚀

File: internal/provider/example_function_test.go (hunk 1 above)
Comment (bflad) [reply internal/provider/example_function_test.go:76]: Thanks! When terraform-plugin-testing v1.7.0 is released we should also be able to clean up and enhance these example tests further. 👍

**Pull request merged**